### ISSUE TITLE
Cache build information between webpack processes

### ISSUE THREAD

**Issue by zkat** [webpack/webpack#250]
The current watch mode for webpack is fairly good at caching load results and general workload between file changes as long as you keep the same webpack process running. The initial build for some projects, on the other hand, can be on the order of more than a minute. This can make the build process very slow if for some reason you need to stop your watcher, or node crashes, or you simply want to restart your connect server with different flags.

It would be nice if webpack were able to persist compilation information between invocations of the node process to the filesystem and reload that cache later -- so ideally, performance between different `$ webpack` calls would be close to how long it takes to `--watch` to recompile when it detects a change.

**Comment by damassi** [comment]
Running into this problem as well.  Once its going its instant, but right now our initial build takes over a minute to warm up.

**Comment by jtangelder** [comment]
This would be really cool.

**Comment by nickdima** [comment]
This could be very useful also for speeding up incremental builds in a continuos deployment setup.
@sokra could you point us in the right direction if we want to help with this?

**Comment by marcello3d** [comment]
Any pointers on where to start to solve this issue?

**Comment by bholloway** [comment]
I took a look at serialising the cache based on a TodoMVC project (babel, sass) that I have lying around and Webpack 1.12.0. Here is what I found.

The `compiler.cache` is provided by `CachePlugin` created by `WebpackOptionsApply`. In `CachePlugin` it is trivial to serialise on the existing `"after-compile"` step and reload the same file (if it exists) in the constructor. Sure it has circular dependencies but [cycle](https://www.npmjs.com/package/cycle) may be used.

However this doesn't get us anywhere. The reason begin that the cache is not a plain data structure. It is composed of Classes that would need to be identified on serialisation and reconstructed on deserialisation. The deserialised object does not implement the requisite methods and the compiler quickly fails on [`cacheModule.needRebuild()`](https://github.com/webpack/webpack/blob/master/lib/Compilation.js#L80).

So how big is this problem? I hacked [this code](https://gist.github.com/bholloway/749be73742318bc14c29) in the [`CachePlugin "after-compile"`](https://github.com/webpack/webpack/blob/master/lib/CachePlugin.js#L40) to see.

I found that all my interesting files (in the small Todo project) were entries in the `cache` hash each with value `instanceof ContextModule`. Noting that `ContextModule` inherits `Module` inherits `DependenciesBlock` which uses `DependenciesBlockVariable`.

Enumerable on these module instances there was also instances of `Parser`, `OriginalSource`, `SourceMapSource` (or Array thereof).

At first glance it is plausible (but suboptimal) to write an external monolithic serialiser/deserialiser that can reconstruct these objects. It would support only this set of Classes but not necessitate any changes to the Webpack codebase.

Likely such and implementation would only be useful for a narrow set of use cases. Given the large number of Classes in the Webpack `/lib` I imagine there would be a quickly growing list of other classes that would need to be included in the monolith.

The alternative is where @sokra brings the **difficult** tag - Much or all of the Webpack `/lib` needs to support some sort of serialisation interface, in order that the cache items may be recursively serialised/ deserialised. This would be the more consistent solution but needs architectural buy-in.

Please let me know if any of this is misinformed. I'm interested to hear other thoughts on this.

I would like to try the monolithic serialiser. Obviously as its author it would be suited to my use case, but also because it might inform a more general solution.

If you know of any other Classes that would need inclusion (or you can discover them by running a similar test) then please let me know. But my hope is that the full cache would not need to be reinstated to get some performance benefit.

**Comment by bholloway** [comment]
Ok, so I have a really naive [implementation](https://www.npmjs.com/package/persistent-cache-webpack-plugin) of the monolithic serialisation. Overall... interesting but currently unusable.

There were **lots** of classes in the `compiler.cache` that I initially didn't see. But thankfully I was able to use the [`require.cache`](https://github.com/bholloway/persistent-cache-webpack-plugin/blob/master/lib/application-classes.js) to discover pretty much every Class and @sokra unintentionally did an absolutely beautiful job of structuring all the classes so that they could be easily reconstituted.

Without optimisation serialisation is **2 orders of magnitude in the red**, which doesn't bode well. Deserialisation is quite fast, leading me to believe it is the class discovery hurting the serialisation. In the short term I added a `persist:false` option to turn off serialisation and just use any existing cache file.

Even under these conditions **performance gain is only 10%** on a medium project. For [very small projects](https://github.com/angularity/angularity-todo-es5/tree/experiment/webpack) this is completely offset by the time to deserialise the cache. At this stage its pretty amazing it produces workable output at all. But what we are all expecting is incremental-size compile times. So I have to conclude that right not **it is not useful**.

Some learning:
* The cache is much more complex than I expected from my experience working with the Browserify cache. I am sure this is essential complexity but I'm hoping we can be selective in what we persist. This really needs some expert knowledge.

* Possibly doesn't need an architectural solution due to the clean nature of @sokra's classes. Notwithstanding the possibility that the whole concept just doesn't work.

* Currently writing JSON to disk. I'm seeing 200MB cache file on a medium size project with built size around 4MB. Binary will help but how much (?).

* With lots of copies of Webpack code it was difficult to detect Classes from the correct paths, making `require.cache` essential. I need to re-confirm this wasn't something stupid I was doing during development which is now fixed. I'd like to [require-dir](https://www.npmjs.com/package/require-dir) on the Webpack `/lib` folder instead.

* The small project mentioned above builds correctly but that is no guarantee for other use cases. I don't have any medium/large projects in a stable state (migrating from Browserify) but I suspect a problem with at least one. All are AngularJS projects. It would be good if others can test their use-case to see where things stand.

I've spent a couple of days on this and need to let it sit for a while. But if it crashes for you certainly [raise an issue](https://github.com/bholloway/persistent-cache-webpack-plugin/issues) and I will try to look at it right away.

This is such a missing link it would be awesome to crack it. If anyone has feedback or PR that can move this plugin forward please let me know.

**Comment by marcello3d** [comment]
@bholloway Nice work and writeup! Apologies if this is obvious, but how much of the 200MB is duplicated? Are the same references being serialized multiple times? You might also be running into memory pressure/gc just from serializing/deserializing the entire thing into memory at once (as opposed to streaming/multiple cache files).

**Comment by bholloway** [comment]
Running an amended version of Douglas Crawford's decycle to remove
duplicates and circular references.

Forgot to mention that the emit phase is running for reach compiler. Need
to find an async  phase for overall completion and act just once.

**Comment by sokra** [comment]
@bholloway pretty awsome work. 

Nice idea of using require.cache for serialization. I initally thought about adding a static property to each class storing `__filename` of the module.

On of the reasons why the cache is so big is propably that modules are stored into the cache while still connected in the graph. They are disconnected when they are read from the cache (search from "disconnect" in the webpack source). I'll try to change this in webpack (may reduce memory usage too) and check how much effect does it have on the serialization.

**Comment by markfinger** [comment]
I tried a similar approach as @bholloway and found the serialization/deserialization costs were too high, it's really nice to see that the problem wasn't as insurmountable as it appeared to me.

I ended up wrapping webpack, caching it's output, and adding a dozen or so lines of cache validation logic (checking mtimes on dependencies and emitted assets, package versions, etc). For our purposes (automated deploys with remote builds) this dropped webpack's build times from ~30 seconds to < 1 second - most of which is spent booting the node process, loading the modules and cache, and hitting the file system.

The same wrapper is used on a bunch of local dev machines as well. We serve from the cache, while spawning a watching compiler in the background. Once the compiler was ready, the cache would delegate to the compiler, and the compiler would keep the cache populated.

**Comment by markfinger** [comment]
Probably repeating @bholloway points, but the blockers that I can remember encountering when trying to add a persistent cache to webpack were:

- Cyclical structures in both the cache and the state
- Mixtures of simple objects and instances with constructers+prototypes

The cyclical structures ensured that you had to walk the entire tree to find any cycles, which is slow on both ends of the serializer.

The objects with constructors and prototypes ensured that you would need a large amount of domain knowledge to cover all the use cases in both webpack and the ecosystem.

If those two issues were solved, performant caching would be pretty viable.

The last thing I remember considering was adding an interface for plugins and loaders to populate and read from the cache.

**Comment by mzgoddard** [comment]
I've been looking at this too. Here's my current implementation. https://github.com/mzgoddard/webpack-cache-module-plugin

I dove into webpack's source while trying my hand at it so far. I realized the cyclical parts of the cache are caused by Compilation handling built modules and primarily storing Reasons that point to the module that depends on the newly built module. If a cached module is used the old Reasons are thrown away and new ones are computed during an iterative build.

From reading some other plugins I thought I might use this detail and hard code the members of modules to be serialized. To deserialize them I'd let webpack's normalModuleFactory do a bunch of the lifting and in the factory's module plugin interface wrap the NormalModule in a proxy object that uses the specifically selected serialized members for the first run if the timestamp wasn't out of date for its dependencies.

This works. It works for at least in my generated sandbox. I think the take away either way is to when serializing or deserializing the cache treat the modules as disconnected (https://github.com/webpack/webpack/blob/f7d799adb4dc52c7548cea63f5b08edfcf4476f6/lib/Module.js#L29-L37). As far as I can tell `Module#disconnect` removes the circular references and its essentially the state the modules are in when they are first created and the state Compilation wants when using a CachedModule.

The other thing I noticed, and I'm not sure if it affects @bholloway's work but I think it does, is even with using these cached proxies around the normal modules I saw maybe a 10% improvement in performance. It wasn't after some more poking around and logging timestamps that I noticed the biggest consumer of time was spent resolving locations of dependencies. Implementing a stored UnsafeCache (https://github.com/mzgoddard/webpack-cache-module-plugin/blob/0f09a55b01715c8e70dbc64098a220b6f5c67f68/lib/CachePlugin.js#L80-L97) gave me the performance improvement we're all hoping for from a persisting cache. Compilation when processing modules for dependencies and recursing, building the depended on modules, has to resolve what that dependency is, or more specifically the module factory has to. I thought I knew how the iterative builds avoided this but looking back at webpack's source I'm not as sure. Just thinking about it now I'd guess the CachedInputFileSystem is helping here once primed from the first run.

Freezing and thawing the cache to disk is pretty straight forward. I think caching dependency resolutions or possibly part of the CachedInputFileSystem is where a large win may be waiting but they are less straight forward. One thought is if we could list all attempted filepaths for context and dependency pair, we might be able to watch or check those paths on start to invalidate path resolutions. Another thought is maybe CachedInputFileSystem can be a help here. Maybe if we persist to disk the stats, readlink and readdir info in there and check it on revalidate that on start we could gain a similar impact to UnsafeCache but be ... safer. I imagine though that revalidating info for the CachedInputFileSystem could take a lot of time considering the number of paths involved in file resolving dependencies and loaders.

To reiterate, from what I've found, Reason and Chunk objects and others removed in `disconnect` stored on Modules are the circular references and are not wanted during iterative compilations so they can be safely ignored. Resolving dependencies is very elaborate and takes a lot of time, to approach iterative build times somehow **safely** caching dependency path resolutions or CachedInputFileSystem is likely a needed step.

Hope this helps.

**Comment by Globegitter** [comment]
@sokra Is there anything in webpack 2 that would make caching to disk easier?

**Comment by markfinger** [comment]
**TLDR:** persistent caching gets a lot easier when it's low-level and granular.

------------------------------------------------------------------------------------------------------

I took some time over the end of year break to do a [greenfield reimplementation](https://github.com/markfinger/unfort) of a subset of webpack's features. It's reasonably functional and I'm using it currently in dev environments on some personal and work things. I do want to clarify that this is not intended as a webpack replacement, it's foremost a research project that I used to investigate and experiment within the problem space of building for the web.

One of the problems that I focused on solving was persistent caching of build data. I went through a couple of iterations before settling on the current approach that balances various trade-offs while providing some significant performance improvements. In one case, a project with an initial build time averaging ~25 seconds drops to ~2 seconds.

**Background**

The persistent cache mechanism that's in there is novel for two reasons - the population and the invalidation mechanisms.

Cache data is populated by various jobs in the pipeline that annotate per-file objects with data that is persisted and - once the process reboots - then rehydrated on-demand as each file is run through the pipeline again. If cached data is available, jobs can simply return the annotation on the object, rather than recalculate anything. To prevent delaying the build process, cache writes occur after the build has been emitted.

Invalidation occurs at two levels: all data is [namespaced](https://github.com/markfinger/env-hash) during the process's boot and the pipeline contributes a [unique cache key](https://github.com/markfinger/unfort/blob/cff4f3a9c20ce28f5f226c50d943026a3ba31980/src/jobs.js#L114-L128) for each file. The namespace enables high-level invalidation of the cached data, it's mostly used to discard data when changes are made to build scripts, node_modules, package.json, etc. The cache key enables low-level invalidation which is contextual to the specific file itself and is composed of the path, mtime and (in some cases, to get around OS time accuracy issues) a hash of the textual content.

**Lessons learnt**

- Storing all data in a single json file is (in some cases) the most performant approach, but it'll lock up your process for a while as JSON.parse or JSON.stringify generates stuff. With a single file, you also need to invalidate data more regularly otherwise the (de)seralization costs grows rapidly. We store each file's data in a distinct file, while there are some IO overheads that this introduces, it helps to reduce blocking the event loop and removes most of the need to invalidate data from the cache.
- Invalidation is a pain and you'll never get it perfect, so make it easy to nuke the cache or turn it off. There are endless edge-cases that you can't reasonably detect.
- [Mock caches](https://github.com/markfinger/kv-cache/blob/c0ac52806a26492fa64383bad0844af2a92be55c/src/mock-cache.js) are useful for testing and profiling code-paths without having to deal with integration issues.
- Leave the cache at an extremely low and granular level so that subsystems can override or ignore it. It's tempting to simply skip jobs if you have cached data that you think is probably correct, but there is a lot of domain knowledge in each part of a build tool and you'll want to enable each part to control how their data is handled. For example: we cache dependency resolution with [varying degrees of aggression](https://github.com/markfinger/unfort/blob/cff4f3a9c20ce28f5f226c50d943026a3ba31980/src/jobs.js#L444-L469) based on the location of a file, in some cases we want to cache everything, in other cases we only want to cache package dependencies - hence, the job still needs to run so that it can decide how much of the data it wants to reuse or recompute.
- Only cache the most expensive data. Most of the data associated with the build can be regenerated in a fraction of a second and hence, there's no need to preserve it as it'll only add to the re-hydration overhead and overly complicate your codebase. Code generation, sourcemaps, and dependency resolution - these are the really expensive parts that you don't want to recompute unless necessary.
- Don't cache large data structures such as ASTs, the (de)serialization cost is enormous. You'll only want to cache the final outputs, such as code strings, sourcemaps or dependency lists.

**Musings/recommendations for webpack**

A persistent caching system comparable to what we've implemented could be added to webpack reasonably easily. I'd strongly recommend making small, incremental steps with it and profiling code to determine the appropriate tasks to cache.

As a starting point, loaders might be an appropriate place. Assuming that the loader interface was extended to add support for caching, they could opt-in to persistent caching of their computed data via an interface like `this.persistentCache()` which accepts a callback or returns a promise resolving to the data. If either the cache or data is not available, they can compute as before, and if the cache is available, populate it before exiting.

By adding opt-in persistent caching support as a low-level functionality for specific parts of the process, build-time could be steadily reduced without adding too much complexity to webpack's core. By making it an opt-in process, it's backwards compatible and is a feature that newer versions of loaders or plugins can take advantage of.

With regards to cache-invalidation, it gets a little bit more complicated, but you could take a comparable approach to what we have taken, eg: high-level and low-level cache invalidation. A start for a high-level scan would be checks on the config file or a hash of the config object, combined with a superficial check on the root of the node_modules dep tree. I extracted the mechanism that we use for high-level scans as [env-hash](https://github.com/markfinger/env-hash), if it's of any assistance. For low-level invalidation, the path, mtime and content hashes would probably be fine.

**Comment by jhnns** [comment]
Thx @markfinger for sharing your insights. :+1:

[I've mentioned your project in a SO answer](https://stackoverflow.com/questions/36329964/will-webpack-adopt-an-ast-as-a-module-input/36365084#36365084).

**Comment by mzgoddard** [comment]
The last few weeks I gave trying to cache webpack builds another look. Since my other post on this issue I've learned more about building complex webpack plugins. That lead me to my current approach which I think is feasible for at least the iterative development use case.

My current work is up on github and npm at https://github.com/mzgoddard/hard-source-webpack-plugin and https://www.npmjs.com/package/hard-source-webpack-plugin respectively.

This plugin doesn't try to freeze and thaw the cache that Compilation and CachePlugin operate on. Instead it stores the rendered source of a module in a **hard** non-mutable shape and its necessary metadata. It invalidates a cached hard source version of a module by not returning a hard source module and instead letting the normal build process occur.

My last disk caching attempt was incredibly fragile. Small missing unconsidered details led to broken builds. Working on that version I was constantly discovered unconsidered elements. I didn't feel confident about it and had no desire to maintain it.

Comparing my experience with hard source to my last attempt, its been drastically different. Every thing I have missed and fixed has been much easier to solve and the number of missed details has been very manageable. Most of the details relate to dependencies. While the hard source drops the use of dependencies for secondary renders of the module, the module still needs the dependencies so webpack can discover the module graph and construct the wrapping around a module that isn't stored in the hard source. What has made this manageable to me is reducing the set of dependencies to a smaller set of custom dependencies than needed to originally render the module.

The only commonality hard source holds with my last attempt is a resolve cache. Like last time I still find caching path resolutions to be important for performance gains. Its not as impactful as the last attempt since much more savings are made with the modules. But still very noticeable.

I hadn't checked this issue before taking this shot at caching so its really refreshing to see that my approach and plans mirrors points made by @markfinger. Like on invalidation. Definitely going to use `env-hash` for assisting in namespace invalidation.

The HardSource plugin also sees a magnitude difference in performance. I haven't been able to test on a large project yet, but testing on two smaller personal projects I've seen times for a ~5s project drop down to ~500ms and a ~8s project drop down to ~1s. Those projects respectively have ~600 and ~400 modules. I already know how to improve the plugin for that second case to drop it into a magnitude difference. (The difference for the latter is a lot of large binary assets that the plugin handles in a very inefficient format at the moment.) I haven't analyzed it enough but at this point I suspect the time left is spent building the module graph, "sealing" the compilation, rendering the chunks, and emitting assets. (Naturally I'm curious about investigating if the plugin can assist with building the module graph and emitting assets. The others too, I just figure those would be easier to bring about an improvement.)

I didn't expect to be so excited to share this and hope it can help lift some of the pain of long webpack build times. Hopefully this can help contribute to any further iterations to building webpack iteratively as well. :smile:

**Comment by markfinger** [comment]
@mzgoddard that looks really cool. I didn't dig too deep into the source, but you're definitely showing a lot of experience with webpack and its internals.

Hmm, since my last post in here, I've continued tinkering and working on some ideas:-

*Cache storage/retrieval*

One of the long standing issues I've found with persistent caching is how to juggle IO vs CPU bottlenecks. Specifically, it's faster to read one file and deserialize the JSON, than it is to do the same for multiple files. However, the larger the JSON data, the longer you'll lock the CPU up during the deserialization process.

I did [some perf tests](https://github.com/markfinger/node-persistent-cache-performance-test) a while back to try and get an idea of the scope and constraints. I didn't get too many clear takeaways, which I suspect is mostly down to modern operating systems having amazingly fast caches.

So, from a mixture of anecdotal tests and past experience, my current plan is to use a [sqlite3 db](https://github.com/mapbox/node-sqlite3). Sqlite is portable, has some amazingly fast reads, and handles the bookkeeping side of file offsets, so that you can avoid issues surrounding the cost of deserializing stale data. Additionally, using a single sqlite db file avoids the `EMFILE` errors that my previous caching mechanism eventually hit.

*Invalidation*

I've been exploring an idea of using a [file system abstraction](https://github.com/markfinger/unfort/tree/master/file_system) to create a dependency graph of each compilation unit.

The current manifestation is that each call to read or stat a file is passed through a file cache that tracks the outcomes from the calls, so that when a file watcher signals a change (or a file add/remove), a set of traps can be triggered that will invalidate any units with stale dependencies. 

Invalidation is also trivial during startup as you can replay the previous build's FS calls and then diff the results of `isFile`, `mtime`, and text hash checks.

While the concept does rely on some extra overhead, I'm particularly hopeful that it'll solve some long-standing complications and edge-cases that arise from caching dependency resolution.

In any case, I'm still a couple of weeks off from wrapping up the project, but I'll try to remember to do a brain dump here once I've gotten past the usual teething problems.

**Comment by markfinger** [comment]
@mzgoddard once your plugin has stabilised, would you able to produce a little spec detailing the hooks and data structures that your plugin depends on? 

My hope is that we may be able to work towards integrating your work into core or, at the least, defining some explicit APIs.

From a little dig through your source, it looked like there was a fair dependency on the implementation and structure of webpack's internals. If those hooks can be explicitly provided, it may pave the way for a robust implementation that the rest of the ecosystem can integrate with.

@amireh may also be able to shed some light based on experience with [happypack](https://github.com/amireh/happypack). 

Possibly related: https://github.com/amireh/happypack/issues/52

**Comment by mzgoddard** [comment]
Gonna respond in order of your posts.

---

Thanks. I'm really happy that my webpack knowledge is getting used this way.

---

You make a really great point on storing and retrieving the cache. Reading and writing out does seem very nuanced. I took a stab at it last night with leveldb as a first shot. Mostly familiarity but figured its stream mechanism and log style writing fits the module cache's use case currently.

I am curious to try sqlite as well. Eventually it'll probably be done through plugins. I figure someone will eventually write a memcache like micro server to hold the cache in memory inbetween runs if it provides further performance gains.

The db style cache does provide some benefit (from my very anecdotal numbers). It led me to parallelize more asset and resolve cache at the same time and that definitely helped too.

I need to find an approach for assets to avoid `EMFILE`. I suppose that'd be just limiting the number of active ops.

---

Your file system abstraction approach for invalidation is really exciting. I look forward to seeing your results.

---

You're reading my mind. I'm definitely looking to expose hooks once I get enough tests to see what the custom version looks like that fulfill said tests. Some obvious hooks will be storing and retrieving the cache, invalidating, serializing and deserializing dependencies / assets / modules.

Whatever we can get into webpack would be great. At this point one API I'd like to see is a way to get the devtool object webpack builds. webpack builds an object that is consumed by all of the `Source`'s source map methods (map, node, listMap, sourceAndMap) but there isn't a clear public location to get this object currently. The plugin could monkey patch those methods on the original NormalModules to hijack the info but I'd rather copy the devtool object creation than do that. Having that object from a public compiler member would also help source map support in loaders reducing source map work done to what will be consumed by webpack.

I'm not yet sure if it needs new access point in webpack 2 but I need to revisit `HardSource`'s harmony module support. webpack 2's tree shaking leads to rendering changes in depended on modules. To support that HardSource needs a way to invalidate a module before Compilation's seal and instruct it to factory create a new Module so that the HardSource provided CacheModule is replaced by the normal build process. I haven't dug into this problem yet to confirm if this would need a new hook or not.

---

I have done some tiny reading on happypack. My high level understanding is it parallelizes and caches work done by loaders instead of the modules. Bet there are ways for our two plugins to work together for mutual benefit. Anything @amireh can add would be awesome.

**Comment by amireh** [comment]
@mzgoddard This is _so so so_ beautiful. Your solution is so synergetic with HappyPack that I think I may have just hit happiness Rank 5.

TL;DR - `HappyPack` accelerates the initial build while `hard-source` takes over the initial build and caches it. Successive builds don't even need HappyPack or its cache anymore and are done in, as you put it, an order of magnitude shorter time.

Doing a quick test on our application at work took the (re)build time down from **90s** to 29s, then 29s to **4.5s**. It's just beautiful.

I don't really know how much I can add to this conversation though... HappyPack is really not as elaborate as `hard-source` or anything. The latter solves a much harder solution and in my opinion is a great fit to be in webpack core.

Anyway, I'll try to add my bits into what HappyPack tries to solve, how it works, and its limitations (which IMHO forbid it from being integrated into webpack core.)

> Invalidation occurs at two levels: all data is namespaced during the  process's boot and the pipeline contributes a unique cache key for each file.

HappyPack adopts a somewhat similar approach but it doesn't even try to be smart about what you dub as namespacing; instead it exposes a configuration hook for defining the cache's "context" (namespace, if you will) which has to be a JSON-serializable object and delegates the responsibility to the user to tell it when to drop its cache.

One could, for example, use `env-hash` to populate this item.

At the file layer, it uses the file path as well as the mtime but no hashing of content - although that's also possible through another configuration hook.

> Only cache the most expensive data.

This is what I realized after mucking around with webpack internals trying to generate a (de)hydratable version of the compilation state; the _bulk_ of the time seemed to be spent on applying loader transformations.

I'll talk more about that below, which is really what HappyPack tries to improve.

**On applying loaders...**

What I noticed was that webpack was being CPU bound during that phase, mostly within babel transformations and node's event loop wasn't helping out much there. So what HappyPack ended up doing was spawning several processes that try to apply as many transformations as possible "in parallel" (by letting the OS figure things out) and send the results back upstream to webpack's compiler.

This process had to be transparent obviously, and the difficulty in doing that was getting loaders to be applied in a different execution environment than the one webpack compiler is running in.

The mechanism by which HappyPack tries to solve that is by providing thin proxy layers around the `LoaderContext` and `Compiler` interfaces to the loaders; these proxy layers accept requests from loaders running in background processes and forward them to the respective instances running in the foreground (owned by webpack) and then pass the results back.

The limitations of that approach are:

1. no support for any synchronous API (like `this.resolve()`)
2. no support for any operation that relies on webpack internal objects (like
   defining plugins during a loader application, which some loaders like
   TypeScript's need to do in order to provide program-level reporting...)

**What I wish webpack did (or didn't) do**

I really wished webpack didn't choose to expose those `this._compiler` and `this._compilation` objects to loaders as they are impossible to share in a distributed environment, and sadly some loader implementations (namely, TypeScript's) mainly revolve around their utilization (even though they're documented as "hacky access") which put HappyPack in a tight corner.

I had already mentioned this to @jhnns in https://github.com/amireh/happypack/issues/52 and it was one of the reasons I didn't consider presenting something like HappyPack to webpack's core.

The other thing is it would've been nice if webpack could modularize the routines around application of loaders so that a plugin like HappyPack can reuse the implementation instead of rewriting it (this is a heavy maintenance burden really). See https://github.com/amireh/happypack/blob/master/lib/applyLoaders.js for an implementation that can be reused by both webpack and HappyPack.

**What I would like to do with this...**

Is take `hard-source`'s cache and store it somewhere everytime our `master` branch is updated, then have a git post-pull hook download that artifact (taking the target platform in mind) and reuse it to skip the initial build!

The ultimate experience to me would be to relieve other developers of the  initial build altogether unless they're working with the build itself.

**Detailed benchmarks**

The following is for a single bundle with 2279 modules.

Elapsed (ms) | Happy? | Hard-Sourced? | DLLs?
------------ | ------ | ------------- | -----
95451*       | No     | No            | No
4155         |        | Yes           |    
32578        | Yes    | No            | 
4354         |        | Yes           |
90216        | No     | No            | Yes
29077        | Yes    |               |
4511         |        | Yes           | 

For this project I'll be substituting happypack's cache entirely with `hard-source` as it's no longer necessary.

* I find this figure to be strange because normally I get much lower times (~60s without happypack and 10-11s with happypack), there may be something up with my box at the moment.

**Comment by markfinger** [comment]
@mzgoddard 

> I took a stab at it last night with leveldb as a first shot. Mostly familiarity but figured its stream mechanism and log style writing fits the module cache's use case currently.

I took a look at leveldb a while back: https://github.com/markfinger/unfort/issues/116#issuecomment-227311525. sqlite + leveldb seem mostly on par. I faintly remember that leveldb has (or had) patchy support for windows.

> I need to find an approach for assets to avoid EMFILE. I suppose that'd be just limiting the number of active ops.

[graceful-fs](https://github.com/isaacs/node-graceful-fs) might be of use. It introduces a small overhead, but it'll handle the EMFILE issues.

**Comment by markfinger** [comment]
@mzgoddard and @amireh, thank you both for all of your input, and a massive thank you for all your efforts across your projects.

----------------------------------------------

With regard to persistent caching, I suspect that we should give @mzgoddard a few weeks to deal with teething issues, but once everything's settled we can sketch out a proposal and start a dialogue with @sokra to try and implement it in core.

While it may be feasible for `hard-source` to remain an external project, effective and performant caching will always ends up tightly coupled to webpack's implementation details. If `hard-source` is left external, small semver-compatible changes in webpack could easily break it. Additionally, this level of coupling puts pressure on webpack to avoid changing undocumented internals, which is an unreasonable expectation for an already large and complicated project.

**Comment by jhnns** [comment]
Thx for your research and discussion on this topic. I really appreciate it and I'm sure we can use some of the insights to improve webpack's core. I'll bring this up at our next weekly meeting. 👍 

@amireh 

> no support for any synchronous API (like this.resolve())

We will deprecate all synchronous APIs in loaders. They were meant as a lean possibility to also use loaders in a node.js environment (where `require()` calls are synchronous), but it turns out that this leads to very complicated loaders (and is not possible always). The `target: node` approach is probably better, although this means that we have to compile node.js code which is a great barrier for many developers.

> no support for any operation that relies on webpack internal objects (like defining plugins during a loader application, which some loaders like TypeScript's need to do in order to provide program-level reporting...)

I don't know what exactly the issue was with the typescript-loader, but we definitely need to resolve their use-case and then remove this access. If it turns out that applying loaders in threads is faster because many compilation tasks a CPU bound, we should go this way.

Personally, I'd like to see some real-world figures where most time is spent in usual projects. I haven't felt that build time was a big pain for me (all my projects finish the initial build after 10s-15s), rebuilds are always under 1s. Could someone with long build times collect these figures? That would be awesome 👍

**Comment by MoOx** [comment]
We just integrate this hard-source-plugin into [Phenomic](http://phenomic.io/) and "docs" build time is now to 13s, compared to previous 33s.

**Comment by amireh** [comment]
@jhnns Check out https://github.com/amireh/happypack/issues/33#issuecomment-221265755 where one of the typescript loader authors explains their use case.. it's definitely valid and I do believe that if we were able to have webpack accommodate it in a more direct approach, everyone will benefit (and then happypack's concept being ported to webpack core becomes a valid possibility.)

> We will deprecate all synchronous APIs in loaders.

This would be really great.

> Personally, I'd like to see some real-world figures where most time is spent in usual projects.

I am also interested in such figures because I believe there are other areas other than the application of loaders that can be optimized (and would help out in the process if I can.)

I've searched for a node profiling tool that would help me figure out where webpack is spending most of its time during build, but I didn't find any that actually worked so I have been relying on webpack's internal profiling mode so far. Do you know of any such tool? (I mean something like the gnuprof chain for C/C++ applications)

**Comment by wkentdag** [comment]
@jhnns check out https://github.com/static-dev/spike-core/issues/156. spike generates static sites using webpack as its core bundler/compiler, and adding a persistent cache feature to webpack core would, for example, directly affect my use case: compiling a massive wordpress site into a static site that I can distribute through a CDN. With a persistent cache, I can run the initial build all day if need be, and then look forward to more normal build times throughout the rest of the development process, since 90% of the static pages i generate on the first run will never get touched again.

this thread has been my first foray into webpack core, so it totally could be too much work to integrate a somewhat tangential feature. btw, this hard-source plugin looks *awesome*, i'm going to try integrating it asap, and that might work well enough for me. just thought I'd throw my two cents in and say that I think this would be a really great feature for anybody working on large webpack projects, including ones that are leveraging other libraries built on top of webpack (eg, spike). thanks to everyboy here for your hard work on the issue 🎉

**Comment by gdborton** [comment]
@jhnns Any tips for breaking down time spent in various parts of the build?

Our project is massive at 451 chunks and 7217 modules, but even building a subset of that leads to long rebuild times in dev.

```
2016-09-02_18:28:16.85185 chunks 12
2016-09-02_18:28:16.85227 modules 1658
2016-09-02_18:28:16.85246 Build complete in 10343ms
```

**Comment by jhnns** [comment]
I think flame graphs could give us better insights on where so much time is spent. I'll try to get some from my current project and post instructions on how to get them (if you do not know it anyway ^^).

We talked about this issue [at our last weekly meeting](https://github.com/webpack/meeting-notes/pull/13/files). We're planing to deprecate all parts of the loader APIs that make parallel compilations impossible (namely sync APIs and `_compiler`/`_compilation`). However, in order to deprecate these we need to provide better alternatives for loaders that had to access these internal objects (namely typescript loaders but probably some others too).

@sokra had the idea of a loader API that allows to hook into different compilation states. Thus, loaders would be more like plugins but still on a per-file basis. He wanted to create a proposal so that we can discuss it with other loader authors.

Since webpack@2 already uses a [separate module](https://github.com/webpack/loader-runner) to execute loaders, we could write a parallel-loader which spawns multiple processes (see our meeting notes for details). The parallel-loader would need to provide a webpack-like loader context and handle all the communication in the background. I think it's a good idea to push this into user-space instead of embedding it into webpack core. This way we can keep the core simple, and parallel compilation is probably not always desired due to the costs of spawning a new process.

**Comment by jhnns** [comment]
The proposal for the loader API is can be found [here](https://github.com/webpack/webpack/issues/2975)

**Comment by amireh** [comment]
> Since webpack@2 already uses a separate module to execute loaders, we could write a parallel-loader which spawns multiple processes (see our meeting notes for details). The parallel-loader would need to provide a webpack-like loader context and handle all the communication in the background.

Honestly, that's exactly what happypack does, although it doesn't use [loader-runner](https://github.com/webpack/loader-runner) since at the time I saw it it didn't actually work for a distributed application (at least in happypack's context). I can't recall the details but it's probably something easily amendable.

**Comment by jhnns** [comment]
> Honestly, that's exactly what happypack does

That's perfect. You could give us valuable feedback if the proposed loader API would make it possible to use happypack rather as a loader than a plugin.

**Comment by abergs** [comment]
@jhnns I'm investigating our webpack 2 build that takes around 2 minutes (.tsx project with about 800 modules).

Could you please describe how to get those flame graphs?

>I think flame graphs could give us better insights on where so much time is spent. I'll try to get some from my current project and post instructions on how to get them (if you do not know it anyway ^^).

**Comment by amireh** [comment]
@abergs run your webpack with node's inspector (a command like the following):

    node --debug-brk --inspect webpack

Then launch your browser to visit the URL that command will show you (make sure you're on node 6+, 7 preferably)

In the browser page, find the Profiles panel and click on "Start Profile", then let the debugger continue (you may need to go back to the Sources panel to do this). Once the run is complete, go to the Profiles panel and click on "Stop Profile" or such, then you'll find the collected profile in a list to the left. That profile will contain the flame graph you're after.

**Comment by abergs** [comment]
Thank you @amireh

**Comment by jpea-imvu** [comment]
Is there filesystem caching yet?

**Comment by abdelbk** [comment]
Has some progress been made to make this feature at the core of webpack ?
The Rails' Asset Pipeline do it perfectly. Is it something worth investigating ?

**Comment by sod** [comment]
The new [webpack-contrib/cache-loader](https://github.com/webpack-contrib/cache-loader) reduces consecutive builds in our environment from 50s to 27s. But you have to ensure that one cache folder isn't used for different configs (like dev/prod).

**Comment by jedwards1211** [comment]
@sod that's an improvement; how much faster do you think it would be if Webpack could load its entire previous state from disk on startup, see what's changed, and then run an incremental rebuild in the exact same way it would in watch mode?  That way the time to invoke webpack again after a single file change would be reduced to just

```
<time to load previous state from disk> + 
<time to determine which files have changed> +
<time to run incremental rebuild for the files that changed>
```
In my environment incremental rebuilds take on the order of 10s, and I would hope that loading the previous state and determining which files have changed take far less time.

**Comment by MagicDuck** [comment]
I agree with @jedwards1211 
We have like 8000 modules in our build and most of the startup time seems to be taken by reading the files and doing the acorn parsing to determine deps. We are already using a babel cache so cache-loader would not help.
I am willing to pitch in and help but hoping somebody with more experience with the codebase/problem space could take the lead or at least give a direction to get started. I think this feature would make a lot of people very happy and help alleviate performance issues.
@sokra @mzgoddard @markfinger

**Comment by jedwards1211** [comment]
> We have like 8000 modules in our build
Hot damn, I wonder if that's a record! 😅

**Comment by MagicDuck** [comment]
That's what you get when legacy code moves to webpack, lol. It's gotten slightly smaller now, but for the longest time we had a massive hairball chunk about 50mb in size, for which i had to write a custom plugin emulating @sokra's agressive splitting plugin to split it into slices that get loaded in parallel. Good times... 😂

**Comment by filipesilva** [comment]
I was running some tests on how `hard-source-webpack-plugin` could be added to the Angular CLI pipeline and the results seem promising.

It does seem fairly sensitive to Webpack internals though. Since it relies on serializing Webpack sources any change in them can possibly break the plugin.

This unfortunately means that it will necessarily lag behind latest Webpack stable its changes are integrated into `hard-source-webpack-plugin`. Webpack 4 also seems to be bringing in a fair number of performance improvements which will probably affect sources and thus need to be integrated into the plugin.

This makes me think that the only way a plugin such as `hard-source-webpack-plugin` could be stable is if it's integrated directly in Webpack, or has access to a stable API (e.g sources supporting serialization). Maybe the second option would provide plugin authors with more options.

**Comment by evilebottnawi** [comment]
It is very old issue and duplicate more detailed issue https://github.com/webpack/webpack/issues/6527. Please leave you opinions and how we can solve this in https://github.com/webpack/webpack/issues/6527 issue. Thanks!
